python/ray/serve/_private/metrics_utils.py

```diff
@@ -307,71 +307,178 @@ def aggregate_avg(
         return self._aggregate_reduce(keys, statistics.mean)
 
 
-def _bucket_latest_by_window(
-    series: List[TimeStampedValue],
-    start: float,
-    window_s: float,
-) -> Dict[int, float]:
+def time_weighted_average(
+    step_series: List[TimeStampedValue],
+    window_start: Optional[float] = None,
+    window_end: Optional[float] = None,
+) -> Optional[float]:
     """
-    Map each window index -> latest value seen in that window.
-    Assumes series is sorted by timestamp ascending.
-    """
-    buckets: Dict[int, float] = {}
-    for p in series:
-        w = int((p.timestamp - start) // window_s)
-        buckets[w] = p.value  # overwrite keeps the latest within the window
-    return buckets
+    Compute time-weighted average of a step function over a time interval.
 
+    Args:
+        step_series: Step function as list of (timestamp, value) points, sorted by time.
+            Values are right-continuous (constant until next change).
+        window_start: Start of averaging window (inclusive). If None, uses the start of the series.
+        window_end: End of averaging window (exclusive). If None, uses the end of the series.
 
-def _merge_two_timeseries(
-    t1: List[TimeStampedValue], t2: List[TimeStampedValue], window_s: float
+    Returns:
+        Time-weighted average over the interval, or None if no data overlaps.
+    """
+    if not step_series:
+        return None
+
+    # Handle None values by using full timeseries bounds
+    if window_start is None:
+        window_start = step_series[0].timestamp
+    if window_end is None:
+        # Use timestamp after the last point to include the final segment
+        window_end = step_series[-1].timestamp + 1.0
+
+    if window_end <= window_start:
+        return None
+
+    total_weighted_value = 0.0
+    total_duration = 0.0
+
+    # Find the value at window_start (LOCF)
+    current_value = 0.0  # Default if no data before window_start
+    for point in step_series:
+        if point.timestamp <= window_start:
+            current_value = point.value
+        else:
+            break
+
+    current_time = window_start
+
+    # Process each segment that overlaps with the window
+    for point in step_series:
+        if point.timestamp <= window_start:
```

Comment:
since window_start is inclusive
```suggestion
        if point.timestamp < window_start:
```